staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,223 @@
+// only in cases where the window of events in the underlying
+// source can change over time - i.e. for watchCache circular
+// buffer. When the circular buffer is full and an event needs
+// to be popped off, watchCache::startIndex is incremented. In
+// this case, an interval tracking a window of events is valid
+// as long as
+//	watchCacheInterval::startIndex >= watchCache::startIndex.
+// When the above condition becomes false, the interval is no
+// longer valid and should not retrieve and serve new events
+// from the underlying source.
+type watchCacheInterval struct {
+	// startIndex denotes the starting point
+	// of interval being considered, the index
+	// value is the index in the actual watch
+	// cache, however this should be used modulo
+	// the capacity of the cache whenever being
+	// used to index the circular buffer.
+	startIndex int
+
+	// endIndex denotes the end of the interval
+	// being considered, the index value is the
+	// index in the actual watch cache, however
+	// this should be used modulo the capacity
+	// of the cache whenever being used to index
+	// the circular buffer.
+	endIndex int
+
+	// indexer takes an index and returns the corresponding
+	// watchCacheEvent from an underlying watch cache.
+	indexer indexerFunc
+
+	// indexValidator is used to check if a given
+	// index is still valid perspective. If it is
+	// deemed that the index is not valid, then this
+	// interval can no longer be used to serve events.
+	//
+	// An interval is invalid when its startIndex
+	// no longer coincides with the startIndex of
+	// underlying watchCache and therefore does not
+	// have meaning for cases where the events are
+	// served from the underlying store instead of
+	// the watchCache (rv = 0).
+	indexValidator indexValidator
+
+	// buffer holds watchCacheEvents that this interval
+	// returns on a call to Next(). This exists mainly
+	// to reduce acquiring the lock on each invocation
+	// of Next().
+	buffer *watchCacheIntervalBuffer
+
+	// lock effectively guards the valid field.
+	lock sync.Locker
+}
+
+type attrFunc func(runtime.Object) (labels.Set, fields.Set, error)
+type indexerFunc func(int) *watchCacheEvent
+type indexValidator func(int) bool
+
+func newCacheInterval(startIndex, endIndex int, indexer indexerFunc, indexValidator indexValidator, locker sync.Locker) *watchCacheInterval {
+	return &watchCacheInterval{
+		startIndex:     startIndex,
+		endIndex:       endIndex,
+		indexer:        indexer,
+		indexValidator: indexValidator,
+		buffer:         &watchCacheIntervalBuffer{buffer: make([]*watchCacheEvent, bufferSize)},
+		lock:           locker,
+	}
+}
+
+// newCacheIntervalFromStore is meant to handle the case of rv=0, such that the events
+// returned by Next() need to be events from a List() done on the underlying store of
+// the watch cache.
+func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc, locker sync.Locker) (*watchCacheInterval, error) {
```

Comment:
We shouldn't need to pass locker here at all - it conceptually shouldn't be needed.